### ISSUE TITLE
lint `if let Some(..) = .. {} else { None }`, which should use `and_then`

### ISSUE THREAD

**Issue by oli-obk** [Manishearth/rust-clippy#1856]
If the body of the `Some` path contains return or break with a label outside the `if let` and the entire statement isn't the last one in a function then this shouldn't lint, since the code paths can't be simplified trivially.

An example in clippy is https://github.com/Manishearth/rust-clippy/blob/master/clippy_lints/src/utils/mod.rs#L265

**Comment by killercup** [comment]
Add a special case if this is the last statement in a block:

I'd actually rewrite [your clippy example](https://github.com/Manishearth/rust-clippy/blob/6a2525ccb84179d00991fdebf19084a976d676f3/clippy_lints/src/utils/mod.rs#L265) to

```rust
let krate = crates.iter().find(…);
let krate = if let Some(krate) = krate { krate } else { return None; };
```

(Or, use a handy `ok_or_return!` macro if I write this ≥3 times in a file. Or, when `?` for Options is stable, use that.)

**Comment by oli-obk** [comment]
Yea, lets wait for `?`
